Assets/Scripts/Menu/UIManager.cs

```diff
@@ -1,29 +1,127 @@
-﻿using UnityEngine;
+﻿using TMPro;
+using UnityEngine;
+using UnityEngine.UI;
 
 public class UIManager : MonoBehaviour
 {
-    private TankUpgradePanel[] factoryPanels;
+    public TextMeshProUGUI moneyText;
+    public TextMeshProUGUI totalTimeText;
+    public Shop shopPanel;
+    
+    private GameManager gameManager;
+    private TankUpgradePanel[] upgradePanels;
+    private Button[] upgradeButtons;
+    private Button[] shopButtons;
+    private TankUpgradePanel activeUpgradePanel;
+
+    private GameObject[] prefabs;
+
+    private void Start()
+    {
+        gameManager = GameManager.instance;
+        shopButtons = shopPanel.GetComponentsInChildren<Button>();
+    }
 
     public void FactorySelected(Factory factory)
     {
         if (factory.upgradePanel.activeSelf)
         {
             factory.upgradePanel.SetActive(false);
+            
+            if (shopButtons == null) return;    //Enable all shop buttons
+            foreach (Button shopButton in shopButtons)
+            {
+                shopButton.interactable = true;
+            }
+            
             return;
         }
         CloseAllActiveUpgradePanels();
-        factory.upgradePanel.SetActive(true);
+        OpenUpgradePanel(factory.upgradePanel);
+    }
+
+    private void OpenUpgradePanel(GameObject upgradePanel)
+    {
+        upgradePanel.SetActive(true);
+        upgradePanel.GetComponent<TankUpgradePanel>().RefreshAllButtons();
+        activeUpgradePanel = upgradePanel.GetComponent<TankUpgradePanel>();
+        CheckUpgradeButtonsBudget(activeUpgradePanel);
+        
+        if (shopButtons == null) return;    //Disable all shop buttons
+        foreach (Button shopButton in shopButtons)
+        {
+            shopButton.interactable = false;
+        }
     }
 
     private void CloseAllActiveUpgradePanels()
     {
-        factoryPanels = GameObject.FindObjectsOfType<TankUpgradePanel>();
-        foreach (TankUpgradePanel tankUpgradePanel in factoryPanels)
+        upgradePanels = GameObject.FindObjectsOfType<TankUpgradePanel>();
+        foreach (TankUpgradePanel tankUpgradePanel in upgradePanels)
         {
             if (tankUpgradePanel.gameObject.activeSelf)
             {
                 tankUpgradePanel.gameObject.SetActive(false);
             }
         }
     }
+
+    public void UpdateMoneyTextUI()
+    {
+        moneyText.text = "$" + PlayerStats.Money.ToString();
+        CheckShopButtonsBudget();
+        CheckUpgradeButtonsBudget(activeUpgradePanel);
+    }
+
+    private void CheckUpgradeButtonsBudget(TankUpgradePanel upgradePanel)
+    {
+        if (!upgradePanel) return;
+        CheckUpgradeButtonBudget(upgradePanel.damageButton, upgradePanel.weaponPrefab.GetUpgradableAttributes().damage);
```

Comment:
👀 we should use a dictionary to get upgradable attributes one by one

Reply:
🤣